lib/time/counter.c

```diff
@@ -0,0 +1,305 @@
+{
+	*seconds = reference_seconds;
+}
+
+static bool reference_is_set(void)
+{
+	return reference_seconds != INT64_MAX;
+}
+
+static void clear_reference(void)
+{
+	reference_seconds = INT64_MAX;
+}
+
+static int get_ticks(uint64_t *ticks)
+{
+	int ret;
+	uint32_t counter_ticks;
+	uint32_t counter_center_bottom_ticks;
+	uint32_t counter_center_top_ticks;
+
+	ret = counter_get_value(counter, &counter_ticks);
+	if (ret < 0) {
+		return ret;
+	}
+
+	counter_center_bottom_ticks = top_ticks >> 2;
+	counter_center_top_ticks = counter_center_bottom_ticks * 3;
+
+	if (counter_ticks < counter_center_bottom_ticks) {
+		get_zero_ticks(ticks);
+		*ticks += counter_ticks;
+	} else if (counter_ticks > counter_center_top_ticks) {
+		get_zero_ticks(ticks);
+		*ticks -= top_ticks;
+		*ticks += counter_ticks;
+	} else {
+		get_center_ticks(ticks);
+		*ticks = center_ticks;
+		*ticks += counter_ticks;
+	}
+
+	return 0;
+}
+
+static void semlock_lock(void)
+{
+	(void)k_sem_take(&semlock, K_FOREVER);
+}
+
+static void semlock_unlock(void)
+{
+	k_sem_give(&semlock);
+}
+
+static void alarm_callback(const struct device *dev,
+			   uint8_t chan_id,
+			   uint32_t ticks,
+			   void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(chan_id);
+	ARG_UNUSED(ticks);
+	ARG_UNUSED(user_data);
+
+	advance_zero_ticks();
+}
+
+static int set_center_alarm(void)
+{
+	const struct counter_alarm_cfg config = {
+		.callback = alarm_callback,
+		.ticks = top_ticks >> 1,
+		.user_data = NULL,
+		.flags = COUNTER_ALARM_CFG_ABSOLUTE,
+	};
+
+	return counter_set_channel_alarm(counter, CENTER_ALARM_ID, &config);
+}
+
+static void cancel_center_alarm(void)
+{
+	(void)counter_cancel_channel_alarm(counter, CENTER_ALARM_ID);
+}
+
+static void top_callback(const struct device *dev, void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
+
+	advance_center_ticks();
+	(void)set_center_alarm();
+}
+
+static int set_counter_top(void)
+{
+	const struct counter_top_cfg config = {
+		.ticks = top_ticks,
+		.callback = top_callback,
+		.user_data = NULL,
+		.flags = 0,
+	};
+
+	return counter_set_top_value(counter, &config);
+}
+
+static int reconfigure_counter(void)
+{
+	int ret;
+
+	cancel_center_alarm();
+
+	ret = set_counter_top();
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = set_center_alarm();
+	if (ret < 0) {
+		return ret;
+	}
+
+	reset_ticks();
+	return ret;
+}
+
+static int configure_counter(void)
+{
+	int ret;
+
+	ret = device_is_ready(counter) ? 0 : -ENODEV;
+	if (ret < 0) {
+		return ret;
+	}
+
+	frequency = counter_get_frequency(counter);
+	top_ticks = counter_get_max_top_value(counter);
+
+	ret = set_counter_top();
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = set_center_alarm();
+	if (ret < 0) {
+		return ret;
+	}
+
+	reset_ticks();
+
+	ret = counter_start(counter);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static void add_ticks_to_seconds(int64_t *seconds, const uint64_t *ticks)
+{
+	*seconds += (int64_t)((*ticks) / frequency);
```

Comment:
You seem to be making implicit rounding assumptions here, please document these assumptions and name the function accordingly. Also document your assumptions about overflow (if any), as well as any other concept relevant to counter-to-time-conversion (see the well-defined concepts and nomenclature in https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731).